### PR TITLE
Improve typing and style and fix some inconsistencies

### PR DESCRIPTION
**Pull request by mariosasko**
This PR:
* improves typing (mostly more consistent use of `typing.Optional`)
* `DatasetDict.cleanup_cache_files` now correctly returns a dict 
* replaces `dict()` with the corresponding literal
* uses `dict_to_copy.copy()` instead of `dict(dict_to_copy)` for shallow copying

### CODE REVIEW

File: src/datasets/features.py [hunk 1]
```diff
@@ -471,15 +473,15 @@ class ClassLabel:
             classes, one per line.
     """
 
-    num_classes: int = None
-    names: List[str] = None
-    names_file: str = None
+    num_classes: Optional[int] = None
+    names: Optional[List[str]] = None
+    names_file: Optional[str] = None
```
Comment (lhoestq) [inline src/datasets/features.py:478]: Currently it's true that these can stay None.
We should disallow that though by doing some changes in the post_init.
This is out of scope for this PR though, so let's keep the Optional for now

File: src/datasets/features.py [hunk 2]
```diff
@@ -471,15 +473,15 @@ class ClassLabel:
...
     id: Optional[str] = None
     # Automatically constructed
     dtype: ClassVar[str] = "int64"
     pa_type: ClassVar[Any] = pa.int64()
-    _str2int: ClassVar[Dict[str, int]] = None
-    _int2str: ClassVar[Dict[int, int]] = None
+    _str2int: ClassVar[Optional[Dict[str, int]]] = None
+    _int2str: ClassVar[Optional[Dict[int, int]]] = None
```
Comment (lhoestq) [inline src/datasets/features.py:484]: same here but out of scope for this PR

File: src/datasets/features.py [hunk 3]
```diff
@@ -667,8 +670,8 @@ class TranslationVariableLanguages:
         }
     """
 
-    languages: List = None
-    num_languages: int = None
+    languages: Optional[List] = None
+    num_languages: Optional[int] = None
```
Comment (lhoestq) [inline src/datasets/features.py:674]: same here but out of scope for this PR

File: src/datasets/utils/version.py [hunk 4]
```diff
@@ -35,10 +36,10 @@ class Version:
...
 
     version_str: str
-    description: str = None
-    major: str = None
-    minor: str = None
-    patch: str = None
+    description: Optional[str] = None
+    major: Optional[str] = None
+    minor: Optional[str] = None
+    patch: Optional[str] = None
```
Comment (lhoestq) [inline src/datasets/utils/version.py:42]:
```suggestion
    major: str = None
    minor: str = None
    patch: str = None
```
These three fields are strings. They're set during post_init

**Review by lhoestq — COMMENTED** [review]
Nice thank you.
This looks all good to me.
I just have one concern about the Version fields major, minor and patch: they're actually always strings so no need to make them optional

File: src/datasets/utils/version.py (hunk 4 above)
Comment (mariosasko) [reply src/datasets/utils/version.py:42]: Checked the souce and it seems these three fields are ints (post_init calls _str_to_version which returns the 3-tuple of ints; the allow_wildcard param of _str_to_version is False by default so '*' is not allowed). I think the correct typing would look like this:

```python
major: int = field(init=False)
minor: int = field(init=False)
patch: int = field(init=False)
```

field(init=False) excludes them from the `__init__` (only version_str is required due to the steps performed in the post_init that are described above).

Another improvement would be to annotate the version_str field with `InitVar[str]`. This would make post_init's implementation cleaner:
```python
def __post_init__(self, version_str):
    self.major, self.minor, self.patch = _str_to_version(version_str) # no longer requires self.version_str
``` 
but with this change dataclasses.fields(cls) can no longer be used in Version.from_dict since it doesn't include info about the class InitVars. So implementation of Version.from_dict would be as follows:
```python
@classmethod
def from_dict(cls, dic):
    field_names = set(f.name for f in cls.__dataclass_fields__.values() if f.init)
    return cls(**{k: v for k, v in dic.items() if k in field_names})
```

**Comment by mariosasko** [comment]
@lhoestq Thanks for the quick review.

File: src/datasets/utils/version.py (hunk 4 above)
Comment (lhoestq) [reply src/datasets/utils/version.py:42]: Thanks for the details.
Let's keep it this way and discuss this in another PR

**Comment by lhoestq** [comment]
I merged master to this branch to re-run the CI before merging :)

**Review by lhoestq — APPROVED** [review]
Thanks :)
